### PR TITLE
 airgeddon: init at 11.01 

### PR DESCRIPTION
**Pull request by PedroHLC**
###### Description of changes

As a red team member, I used to run everything by hand since most of the helpers gave me more headaches than actually automating something. But years ago, I found [airgeddon](https://github.com/v1s1t0r1sh3r3/airgeddon), which is pretty KISS and mostly helps in most simple scenarios. Script kiddies will be happy!

Darwin is not compatible, according to [the author's wiki](https://github.com/v1s1t0r1sh3r3/airgeddon/wiki/Known%20incompatibilities). WSL neither.

Related to https://github.com/NixOS/nixpkgs/issues/81418

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/networking/airgeddon/default.nix
Comment (AtilaSaraiva) [inline pkgs/tools/networking/airgeddon/default.nix:30]:
```suggestion
, reaverwps-t6x # Could be the upstream version too
```

File: pkgs/tools/networking/airgeddon/default.nix
Comment (AtilaSaraiva) [inline pkgs/tools/networking/airgeddon/default.nix:127]:
```suggestion
  meta = with lib {
    description = "Multi-use TUI to audit wireless networks. ";
    homepage = "https://github.com/v1s1t0r1sh3r3/airgeddon";
    license = licenses.gpl3Plus;
    maintainers = with maintainers; [ pedrohlc ];
    platforms = platforms.linux;
  };
}
```

File: pkgs/tools/networking/airgeddon/default.nix
Comment (AtilaSaraiva) [inline pkgs/tools/networking/airgeddon/default.nix:114]: I don't think this is necessary since the license is already specified in the metadata of the derivation, right below.

**Review by AtilaSaraiva — CHANGES_REQUESTED** [review]
Hey Pedro, I'd like to first welcome you to the nixpkgs maintainers team! It is always nice to have more helping hands, and I've been willing to have you help us for a long time. I'm super happy with you first PR.

The PR looks good, although I made some suggestions to make the code just a bit cleaner.

Result of `nixpkgs-review pr 173885` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>airgeddon</li>
  </ul>
</details>

**Comment by PedroHLC** [comment]
Result of `nixpkgs-review pr 173885` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>airgeddon</li>
  </ul>
</details>

**Comment by PedroHLC** [comment]
@AtilaSaraiva suggestions added, rebased, and the package still builds.

File: pkgs/tools/networking/airgeddon/default.nix
Comment (AtilaSaraiva) [inline pkgs/tools/networking/airgeddon/default.nix:37]: What is exactly the reason for this with lib?

File: pkgs/tools/networking/airgeddon/default.nix
Comment (AtilaSaraiva) [inline pkgs/tools/networking/airgeddon/default.nix:107]: You could just have
```
    wrapProgram $out/bin/airgeddon --prefix PATH : ${lib.makeBinPath deps}
```
and I guess the effect would be the same.

File: pkgs/tools/networking/airgeddon/default.nix
Comment (AtilaSaraiva) [inline pkgs/tools/networking/airgeddon/default.nix:119]: Hm I see now, u wanted to use the with lib for those two situations. This is not normally done but it is not wrong, so I guess it is fine.

File: pkgs/tools/networking/airgeddon/default.nix
Comment (AtilaSaraiva) [inline pkgs/tools/networking/airgeddon/default.nix:81]:
```suggestion
  nativeBuildInputs = [ makeWrapper ];
```
This is supposed to only be ran in the build machine, so nativeBuildInputs is the correct choice afaik.

**Review by AtilaSaraiva — COMMENTED** [review]
I left another suggestion pedro, take a look!

File: pkgs/tools/networking/airgeddon/default.nix
Comment (thiagokokada) [inline pkgs/tools/networking/airgeddon/default.nix:103]: This does looks like a good use-case for `resholve.mkDerivation`: https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/misc/shunit2/default.nix

File: pkgs/tools/networking/airgeddon/default.nix
Comment (thiagokokada) [reply pkgs/tools/networking/airgeddon/default.nix:37]: In general we prefer not to use top-level `lib` because this polutes the global namespace. Also `with` is kinda problematic for performance reasons.

File: pkgs/tools/networking/airgeddon/default.nix
Comment (PedroHLC) [reply pkgs/tools/networking/airgeddon/default.nix:143]: I did try, and it helped me find many non-listed dependencies. But this is beyond a simple packager's job. This script generates other scripts in `/tmp` during runtime and launches them in nested tmux/xterm sessions.

Attaching the effort it took for it to launch, in the hope it may help find more dependencies in later updates: https://gist.github.com/PedroHLC/e6cefa4dce6474d27f3a91f4ff929a67

Comment (thiagokokada) [reply pkgs/tools/networking/airgeddon/default.nix:143]: CC @abathur. You may want to take a look at this.

Comment (abathur) [reply pkgs/tools/networking/airgeddon/default.nix:143]: Hehe. I keep a search open, so I'd noticed your comment this morning. I noticed how large airgeddon was and was going to be a bit surprised if there wasn't a blocker in there somewhere.

I assume it'll just make a good addition to the list of ~white-whales, but I'll try to find time to poke at it tonight in case there's something that I know a way past.

@PedroHLC thanks for posting the gist :)

Comment (abathur) [reply pkgs/tools/networking/airgeddon/default.nix:143]: I gave this a try myself tonight and have 3 different thoughts:
1. Thanks for giving resholve a really thorough try, @PedroHLC. This is one of the most ambitious packages a first-time resholve user has attempted, so I'm curious how well the documentation worked. (Feel free to ignore this--I don't really expect more of your time. But, I'd also be happy to chat about it over Matrix, or via the email in my profile?)
2. I think abathur/resholve#78 may make a few bits of this a little cleaner, so I'm inclined to hold up for that before giving it another try. I'll bump it up my priority list.
3. > @PedroHLC: This script generates other scripts in /tmp during runtime and launches them in nested tmux/xterm sessions.

    I'm curious if resholve was causing a specific problem here, and if so--what it looked like?

Comment (PedroHLC) [reply pkgs/tools/networking/airgeddon/default.nix:143]: I'll give you a step-for-step introduction of what I did and what I found, maybe this can help you to help me:

I first took a look at Kokadas' derivation and saw that it was primarily rounded on that "solutions" entry after adding the "mkDerivation". So I searched how to write one of those, which led me to [this README](https://github.com/nixos/nixpkgs/blob/master/pkgs/development/misc/resholve/README.md) that I've used as the source of truth.

My first doubt was, "Should I use the paths relative to input or output?" and well, that was written there, "$out-relative paths".

The second was an observation that I've seen in Kokadas' code. He was using "sh" instead of "bash" as interpreter, which led me to ask him and then discovering bash [includes a posix-mode](https://www.gnu.org/software/bash/manual/html_node/Bash-POSIX-Mode.html).

Then a small try after just adding the "scripts", "interpreter", and the "inputs" I've already known lead me to that regex parsing issue of oilshell, which I solved with the sed `s|^(.+) =~ ([^\$].+) ]]|regexp='\2'; \1 =~ $regexp ]]|` that I based in a commit from you in 2021.

After that I had to deal with a `Can't resolve dynamic argument in 'source'` -- that's because airgeddon sources its scripts with `source "${scriptfolder}${rc_file_name}"` and well, I don't know how to correctly solve these even now :see_no_evil: -- Should I replace this with relative-hardcoded-paths, I know that I can't use absolute-hardcoded-paths -- Kokada pointed me to "`keep.source`" but this didn't help, "`fix`" somehow did the trick for the "strings" and "known_pins", and for the "plugins" in a little hacky way. But I had to leave ".airgeddonrc" out of this and use "`keep`" on it. (very :lost: even now)

At this point, I had already taken a look at the resolve docs. And well, I've started to add all the deps that weren't documented and resholve was finding :trophy: . That lead me to these issues in the way:

- "There is not yet a good way to resolve 'ping' in Nix builds." (Which don't come up with any suggestion, I've added it as an external)

- "timeout" had a "15" where the lore was expecting an "executable argument" (No clue how to say it to skip only this parameter, so I changed it to "cannot", which I'm pretty sure I should not)

- you repeat `"execer"` twice here: https://github.com/abathur/resholve/blob/f3faceb58d731e8d9f0dd244527007cf7a1fbfcc/docs/examples/lore.nix#L14 -- should be "wrapper" right? (This repeats in that README)

- "tmux" is executing its arguments, even worse, airgeddon uses "send-keys" to throw commands in it. (What should I do here? Again I went with "cannot")

- I had to come up with a hacky solution to keep optional dependencies optional, using the "external".

In the end, it ran, but the scripts in heredocs inside it were not using absolute references. They were even using the unpatched shebang. Airgeddon starts running one of these inside tmux and that wasn't finding the executables.

File: pkgs/tools/networking/airgeddon/default.nix
Comment (SuperSandro2000) [inline pkgs/tools/networking/airgeddon/default.nix:155]:
```suggestion
    cp -dr .airgeddonrc known_pins.db language_strings.sh plugins/ "$out/share/airgeddon/"
```
I am pretty sure nix should fix this on its own.

File: pkgs/tools/networking/airgeddon/default.nix
Comment (SuperSandro2000) [inline pkgs/tools/networking/airgeddon/default.nix:82]:
```suggestion
  ] ++ lib.optionals supportWpaWps [
```

File: pkgs/tools/networking/airgeddon/default.nix
Comment (SuperSandro2000) [inline pkgs/tools/networking/airgeddon/default.nix:86]:
```suggestion
  ] ++ lib.optionals supportHashCracking [
```

File: pkgs/tools/networking/airgeddon/default.nix
Comment (SuperSandro2000) [inline pkgs/tools/networking/airgeddon/default.nix:93]:
```suggestion
  ] ++ lib.optionals supportEvilTwin [
```

File: pkgs/tools/networking/airgeddon/default.nix
Comment (SuperSandro2000) [inline pkgs/tools/networking/airgeddon/default.nix:104]:
```suggestion
  ] ++ lib.optionals supportX11 [
```

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 173885` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>airgeddon</li>
  </ul>
</details>

**Review by thiagokokada — APPROVED** [review]

**Pull request merged**
